### PR TITLE
Issue/4571 pn open show pn payload

### LINKED ISSUE #4571: Notifications don’t always open on Android when tapping on a Push Notification
### Expected behavior
When tapping on a WordPress push notification in the device notification center, the full notification content should open inside the app.

### Actual behavior
I tap on the notification and the WordPress app opens on the notifications tab showing the full list of all my read notifications, but the one that I tapped on is not in the list, as if it doesn’t actually exist. 
(The toast "Could not open the notification" can appear for few secs).

If I then wait for a while (anything from a few seconds to a minute or two), the unread notification will appear at the top. 
Sometimes I have to kill the app completely and reopen it to get the notification to be displayed in the list.

### Steps to reproduce the behavior
This seems to happen mostly when I’m on a slower internet connection (mobile data with poor reception or unreliable wifi), so what I’m guessing is happening is that the push notification is received, but the synch with Simperium fails, or it's not started at all.
Btw, sometimes it happened also when I was on fast connections.

##### Tested on Nexus 5x Android 7.0

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4571 

**Important**: @daniloercoli this PR replaces any other pending PRs (i.e. #4609) as this alternative is more elegant. Will close the other PR in favor of this one if you agree (that is I'll leave the other one live until I get your confirmation)

Description: when simperium lags in syncing and we can't find the corresponding Note in the local bucket through its Note id,  this solution implements reading the Note from the Push Notification payload and saving it to the local Simperium Notes bucket to be able to show it to the user.
This implementation makes the fragments agnostic of this change, by capturing the PN payload data right when it is received and saving it to the Note bucket at that very moment.

To test: 

1. comment out the contents of method `startBuckets()`  in file `SimperiumUtils.java`. This will prevent Simperium from actually trying to sync buckets, so we make sure we are not getting info from Simperium as per the purpose of this test.
2. generate a comment on a blog you're an admin of (make the comment with another test user).
3. you should receive a push notification  (which internally carries the full not payload)
4. tap on such notification
5. the app will build the Note out of the PN payload and show the comment contents to the user.
6. also, the user should be able to moderate the comment accordingly

Needs review: @daniloercoli

### CODE REVIEW

**Comment by daniloercoli** [comment]
Really nice work Mario! This approach seems way more cleaner to me.
I'm going to close the other PR and test this new one over the weekend. Will review it on Monday, but at first look it seems OK.

**Comment by daniloercoli** [comment]
Do you think we should merge this against 6.0?

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 1]
```diff
@@ -583,8 +593,53 @@ public Note build(String key, JSONObject properties) {
...
+            // Decompress the payload
+            Inflater decompresser = new Inflater();
+            decompresser.setInput(b64DecodedPayload, 0, b64DecodedPayload.length);
+            byte[] result = new byte[4096]; //max length an Android PN payload can have
+            int resultLength = 0;
+            try {
+                resultLength = decompresser.inflate(result);
+                decompresser.end();
+            } catch (DataFormatException e) {
+                e.printStackTrace();
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: We should log by using `Log.e` method call.

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 2]
```diff
@@ -583,8 +593,53 @@ public Note build(String key, JSONObject properties) {
...
+                    if (jsonObject.has("notes")) {
+                        JSONArray jsonArray = jsonObject.getJSONArray("notes");
+                        if (jsonArray != null && jsonArray.length() == 1) {
+                            jsonObject = jsonArray.getJSONObject(0);
+                        }
+                    }
+                    note = build(noteId, jsonObject);
+
+                } catch (JSONException e) {
+                    e.printStackTrace();
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: Same of above.

File: WordPress/src/main/java/org/wordpress/android/ui/notifications/utils/SimperiumUtils.java [hunk 3]
```diff
@@ -153,6 +160,17 @@ public static boolean hasUnreadNotes() {
         return false;
     }
 
+    public static void saveNote(Note note) {
+        if (note != null) {
+            try {
+                Note object = getNotesBucket().insertObject(note.getId(), Note.Schema.getJSON(note));
+                object.save();
+            } catch (BucketObjectNameInvalid e) {
+                e.printStackTrace();
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/notifications/utils/SimperiumUtils.java]: We should log this error by using our `Log.e` method.

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 4]
```diff
@@ -296,7 +300,9 @@ public void markAsRead() {
             Log.e(TAG, "Unable to update note read property", e);
             return;
         }
-        save();
+
+        if (getBucket() != null)
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: Question: Should not we add braces to this statement?

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 5]
```diff
@@ -583,8 +593,53 @@ public Note build(String key, JSONObject properties) {
...
+            int resultLength = 0;
+            try {
+                resultLength = decompresser.inflate(result);
+                decompresser.end();
+            } catch (DataFormatException e) {
+                e.printStackTrace();
+            }
+
+
+            String out = new String(result, 0, resultLength, StandardCharsets.UTF_8);
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: This call requires API level 19 (current min is 16).

File: WordPress/src/main/java/org/wordpress/android/models/Note.java [hunk 6]
```diff
@@ -583,8 +593,53 @@ public Note build(String key, JSONObject properties) {
...
+            if (out != null) {
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: This condition is always 'true'.

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/models/Note.java]: now that I had to wrap the above line in a try-catch structure, not anymore :)

**Comment by mzorz** [comment]
Ready for a second pass @daniloercoli !

**Comment by daniloercoli** [comment]
:shipit:

**Pull request merged**